crates/block-producer/src/psc.rs

```diff
@@ -736,6 +766,73 @@ async fn median_gte(rpc_client: &RPCClient, timestamp_millis: u64) -> Result<()>
     Ok(())
 }
 
+/// Submit next pending l1 upgrade
+async fn submit_pending_l1_upgrade(ctx: &PSCContext) -> Result<()> {
+    let snap = ctx.store.get_snapshot();
+    let block_number = snap
+        .get_last_confirmed_block_number_hash()
+        .expect("last confirmed")
+        .number()
+        .unpack()
+        + 1;
+    drop(snap);
+
+    // find next l1 upgrade by block number
+    if let Some(l1_upgrade) = ctx
+        .rollup_context()
+        .fork_config
+        .pending_l1_upgrades
+        .iter()
+        .find(|l1_upgrade| l1_upgrade.height == block_number)
+    {
+        let tx: Transaction = l1_upgrade.signed_transaction.clone().into();
+        // update local cell and status
+        ctx.local_cells_manager
+            .lock()
+            .await
+            .apply_tx(&tx.as_reader());
+
+        log::info!(
+            "sending l1 upgrade transaction 0x{}",
+            hex::encode(tx.calc_tx_hash().as_slice())
+        );
+
+        // We can't recover from errors when sending l1 upgrade tx, so we just throw it
+        send_transaction_or_check_inputs(&ctx.rpc_client, &tx).await?;
+        log::info!("tx sent");
+    }
+    Ok(())
+}
+
+/// Confirm next pending l1 upgrade
+async fn confirm_pending_l1_upgrade(ctx: &PSCContext) -> Result<()> {
+    let snap = ctx.store.get_snapshot();
+    let block_number = snap
+        .get_last_confirmed_block_number_hash()
+        .expect("last confirmed")
+        .number()
+        .unpack()
+        + 1;
+    drop(snap);
+
+    // find next l1 upgrade by block number
+    if let Some(l1_upgrade) = ctx
+        .rollup_context()
+        .fork_config
+        .pending_l1_upgrades
+        .iter()
+        .find(|l1_upgrade| l1_upgrade.height == block_number)
+    {
+        let tx: Transaction = l1_upgrade.signed_transaction.clone().into();
+        // We can't recover from errors when confirming l1 upgrade tx,
+        // both deadcell error and unknwown cell is unacceptable, so we just throw it
+        poll_tx_confirmed(&ctx.rpc_client, &tx).await?;
+        log::info!("l1 upgrade tx confirmed");
+        ctx.local_cells_manager.lock().await.confirm_tx(&tx);
+    }
+    Ok(())
```

Comment:
Maybe we should reload `ContractsCellDepManager` when the `upgrade transaction` is confirmed, because the ContractsCellDepManager.deps maybe changed by the `upgrade transaction`.

What do you think?

Reply:
Good catch

Reply:
Fixed